sklearn_numba_dpex/common/kernels.py

```diff
@@ -394,44 +320,52 @@ def partial_sum_reduction(
-        first_work_id = local_work_id == zero_idx
 
-        size = summands.shape[minus_one_idx]
+        sum_axis_size = summands.shape[minus_one_idx]
+        n_work_group_per_row = result.shape[minus_one_idx]
 
-        local_values = dpex.local.array(local_values_size, dtype=dtype)
+        local_work_group_id_in_row = (group_id % n_work_group_per_row)
+        row_idx = group_id // n_work_group_per_row
 
-        first_value_idx = group_id * work_group_size * two_as_a_long
+        first_value_idx = local_work_group_id_in_row * reduction_block_size
         augend_idx = first_value_idx + local_work_id
         addend_idx = first_value_idx + work_group_size + local_work_id
 
+        local_values = dpex.local.array(local_values_size, dtype=dtype)
+
         # Each work item reads two value in global memory and sum it into the local
         # memory
-        if augend_idx >= size:
-            set_col_to_zero(local_values, local_work_id)
-        elif addend_idx >= size:
-            copy_col(summands, augend_idx, local_values, local_work_id)
+        if augend_idx >= sum_axis_size:
+            pass
+        elif addend_idx >= sum_axis_size:
+            local_values[local_work_id] = fused_unary_func_(
+                summands[row_idx, augend_idx]
+            )
         else:
-            add_cols(summands, augend_idx, addend_idx, local_values, local_work_id)
+            local_values[local_work_id] = fused_unary_func_(
+                summands[row_idx, augend_idx]
+            ) + fused_unary_func_(summands[row_idx, addend_idx])
 
         dpex.barrier(dpex.CLK_LOCAL_MEM_FENCE)
         current_n_work_items = work_group_size
         for i in range(local_n_iterations):
             # We discard half of the remaining active work items at each iteration
             current_n_work_items = current_n_work_items // two_as_a_long
-            if local_work_id < current_n_work_items:
-                add_cols_inplace(
-                    local_values,
-                    local_work_id + current_n_work_items,
-                    local_work_id
-                )
+            if (local_work_id < current_n_work_items) and (
+                        first_value_idx + local_work_id + current_n_work_items
+                        ) < sum_axis_size:
+                local_values[local_work_id] += local_values[
+                    local_work_id + current_n_work_items
+                ]
 
             dpex.barrier(dpex.CLK_LOCAL_MEM_FENCE)
 
         # At this point local_values[0] + local_values[1] is equal to the sum of all
         # elements in summands that have been covered by the work group, we write it
         # into global memory
-        if first_work_id:
-            add_first_cols(local_values, result, group_id)
+        if local_work_id == zero_idx:
+            result[row_idx, local_work_group_id_in_row] = (local_values[zero_idx]
+                                                           + local_values[one_idx])
```

Comment:
Does this read a little bit better?
```suggestion
            result[row_idx, local_work_group_id_in_row] = (
                local_values[zero_idx] + local_values[one_idx]
            )
```